lib/metasploit/framework/login_scanner/redis.rb

```diff
@@ -15,12 +15,40 @@ class Redis
         include Metasploit::Framework::LoginScanner::RexSocket
         include Metasploit::Framework::Tcp::Client
 
+        # Required to be able to invoke the scan! method from the included Base module.
+        # We do not use inheritance, so overwriting a method and relying on super does
+        # not work in this case.
+        alias parent_scan! scan!
+
         DEFAULT_PORT          = 6379
         LIKELY_PORTS          = [ DEFAULT_PORT ]
         LIKELY_SERVICE_NAMES  = [ 'redis' ]
         PRIVATE_TYPES         = [ :password ]
         REALM_KEY             = nil
 
+        # Attempt to login with every {Credential credential} in
+        # {#cred_details}, by calling {#attempt_login} once for each.
+        #
+        # If a successful login is found for a user, no more attempts
+        # will be made for that user. If the scanner detects that no
+        # authentication is required, no further attempts will be made
+        # at all.
+        #
+        # @yieldparam result [Result] The {Result} object for each attempt
+        # @yieldreturn [void]
+        # @return [void]
+        def scan!(&block)
+          first_credential = to_enum(:each_credential).first
+          result = attempt_login(first_credential)
```

Comment:
This would potentially lose any surrounding begin/rescue semantics in the super implementation, as well as explicitly retrying the first credential twice. However I think this is not a blocker and will be easy to iterate on for the scope of any future Redis scanner work

Reply:
Can this use `attempt_login` here? Or do we need to use the same semantics as the wrapping module's `requires_password?` implementation?

Reply:
The `attempt_login` is actually the "correct" way in my opinion, as we are doing an actual `AUTH` command that way, and attempting to login for the `default` user with whatever password the credential has set, which will fail when no password is configured, even when sending a blank password.

The question is how long this is sustainable, as Redis clearly labeled this as a backwards compatibility check for older clients, as the command expects an actual username + password combination, but you cannot do that on Redis versions older than V6.

The `requires_password?` method simply sends an `INFO` command, and we assume that authentication is required when an error is received. I'm inclined to follow that approach at this point as this works a bit better with the various versions.

Reply:
I've updated the logic in this class to keep using `AUTH` but handle the case for each Redis version individually.
Tested locally, and wrote an RSpec test to mimic the behavior.